components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth/OAuthServiceTest.java

```diff
@@ -0,0 +1,479 @@
+        String oauthTokenVerifier = "oauth-token-verifier";
+
+        // Create Parameters.
+        Parameters requestParams = new Parameters();
+        requestParams.setOauthConsumerKey("consumer-key");
+        requestParams.setOauthNonce("oauth-nonce");
+        requestParams.setOauthTimeStamp("1506889800000");
+        requestParams.setOauthSignatureMethod("HmacSHA1");
+        requestParams.setBaseString("http://is.com:8080/playground.com");
+        requestParams.setHttpMethod("HTTP-POST");
+        requestParams.setScope("openid");
+        requestParams.setOauthToken("oauth-token");
+        requestParams.setOauthTokenVerifier(oauthTokenVerifier);
+
+        // Create consumer signature.
+        String signature = getConsumerSignature(requestParams, consumerSecret, tokenSecret);
+
+        // Set the created signature to the oAuthConsumer.
+        requestParams.setOauthSignature(signature);
+
+        // Prepare response for OAuthConsumerDAO.getRequestToken().
+        Parameters requestToken = new Parameters();
+        requestToken.setOauthTokenVerifier(oauthTokenVerifier);
+        requestToken.setAuthorizedbyUserName(authorizedSubject);
+
+        prepareForValidateTimestampAndNonce(consumerSecret, false);
+        whenNew(OAuthConsumerDAO.class).withAnyArguments().thenReturn(oAuthConsumerDAO);
+        when(oAuthConsumerDAO.getOAuthConsumerSecret(anyString())).thenReturn(consumerSecret);
+        when(oAuthConsumerDAO.getOAuthTokenSecret(anyString(), anyBoolean())).thenReturn(tokenSecret);
+        when(oAuthConsumerDAO.getRequestToken(anyString())).thenReturn(requestToken);
+
+        OAuthService oAuthService = new OAuthService();
+        Parameters responseParams = oAuthService.getAccessToken(requestParams);
+        assertNotNull(requestParams.getOauthTokenVerifier(), "In the reponseParams should have the tokenVerifier.");
+        assertEquals(responseParams.getOauthTokenVerifier(), requestParams.getOauthTokenVerifier(), "In both response" +
+                " and request parameters OAuthTokenVerifier should be the same.");
+        assertNotNull(responseParams.getAuthorizedbyUserName(), "AuthorizedbyUserName should not be null in the " +
+                "responseParams.");
+        assertTrue(StringUtils.isNotBlank(responseParams.getOauthToken()), "OAuthToken should not be blank in the " +
+                "responseParams.");
+        assertTrue(StringUtils.isNotBlank(responseParams.getOauthTokenSecret()), "OAuthTokenSecret should not be " +
+                "blank in the responseParams.");
+    }
+
+    @Test(dataProvider = "testGetAccessTokenException", expectedExceptions = AuthenticationException.class)
+    public void testGetAccessTokenException(String consumerSecret, boolean isValidSignature) throws Exception {
+
+        String tokenSecret = "token-secret";
+        String oauthTokenVerifier = "oauth-token-verifier";
+
+        // Create Parameters.
+        Parameters requestParams = new Parameters();
```

Comment:
Why not move the request param building to a util method since it is repeated in several places

Reply:
request params are the inputs of each method. In different methods required request parameters are different.
If we are to write an util method we need to have overloaded methods for each and every combination.
In simple terms, we need to have different (overloaded) util methods each and every time request parameters are created.
So this way is better IMO.